### DISCUSSION TITLE
Livery Refit Alternative

### DISCUSSION THREAD

**Discussion by FranticFairy** [OpenTTD/OpenTTD#9904] [Features]
Livery refits, as they currently are, can be a mess at times; it doesn't work well with autoreplace and you cannot refit a full vehicle in one go unless they have the same livery option (under the same 'id', iirc from my earlier GRFs):
https://media.discordapp.net/attachments/477434889508093952/980731835678081074/unknown.png

However, despite various discussions, no real solution has come forth that is as suitable as the current system, despite it's flaws.

However, I believe that a discussion earlier today has led to a possible solution, hence this discussion.

To start with, an extra column when setting company colors, but instead of selecting a color you select a 'livery' from a list of liveries defined by the GRFs you have enabled

https://media.discordapp.net/attachments/477434889508093952/980739979120607252/group_livery_ui_example.png

This sets a group's 'default' livery, and vehicles within this group will use this livery if available. If not available, they fall back to whatever the grf determines to be default.

Next up would be the ability to set liveries individually per wagon/locomotive/RV/etc. This would be done by clicking on the vehicle in question while holding a certain key*

This would then bring up a menu where you get a list of valid liveries for this vehicle specifically, and get to pick which one you want to use for this vehicle.

From there on, a vehicle would pick it's appearance in the following order:
1) If a livery is set for this vehicle specifically, use it. If not:
2) If a livery is set for the group this vehicle is part of, use it if it's valid for this vehicle. If not:
3) Take the default appearance provided by the grf (which can be dependent on year, cargo, etc. Just like now)

This solution would provide support for both convenient group-based liveries, as well as more in-depth choice for those who wish to further customize their trains.

*: My original idea was for this key to be alt, but apparently that's not an option? Another key could be used for this - perhaps CTRL, letting GRF devs decide whether their stuff can flip or change livery (which can also be used for flipping). Please suggest other valid keys.

**Comment by andythenorth** [comment]
There was extended discussion about this on Discord.

TL:DR; probably there will never be a liveries spec that meets the desires of **every** author and **every** player.  But this one would cover a lot of common cases.  Some people will just have to live without ponies.  Into every life a little rain must fall?

**Comment by FranticFairy** [comment]
Couldn't help but come up with a more detailed description of how I envision the implementation of this from the GRF Code side.

To start with, a GRF can define Livery Groups, which show up in the company color list, and Vehicle Liveries, which show up only when changing the livery of an individual vehicle.

If a consist/vehicle belongs to a group with an assigned Livery Group, then that Livery Group becomes available for that consist to do things with just like how you can check for a vehicle's cargo or anything. This way, you can run switches based on the livery of the consist, allowing you to use it to change sprites or do other things. If a livery comes from a different GRF, it will likely be ignored because the consist will not have code written for it - and therefore fallback to the default things set for the consist, unless the author has written code to support this specific Livery Group (there should probably be a consistent name for certain popular Livery Groups so everyone can make use of the same name and ensure compatibility. The way I'm imagining it is that a Livery Group will have an ID like "BRBLUE", and Livery Groups with the same ID get merged into one in the company color screen. This way the list is not cluttered but multiple GRFs can define it).

Beyond that, individual vehicles can also be given their own Vehicle Liveries, similar to how Livery Refits are currently defined - but instead, they only appear when you click on a vehicle in the buymenu, as opposed to refitting the whole consist at once. Again, this livery can then be called upon with switches in the code so that a GRF Author can do what they want with it.

This way, this implementation would preserve the same functionality as what we currently have, while avoiding the issues with autoreplace and cargo refit weirdness. The only situation I can currently think of where this would be a step back is if you want a few consecutive wagons in a bigger consist to be the same livery but different from the rest of the consist - then this system would be a few more clicks on the user's end, since you can't select multiple wagons and apply the same livery option in one click. If you want to refit an entire train, you'd use a Livery Group. If you want to refit individual wagons to different liveries, this implementation requires the same amount of clicks.

**Reply by andythenorth** [reply]
I did a spec for an older idea I had, which put liveries into feature 14.  Might be a starting point.

https://gist.github.com/andythenorth/636908e3498197124782c6daf40de6ae

Lots in there I don't like (dates etc).  But eh.

**Comment by LC-Zorg** [comment]
I like this idea. :) However, I would like to mention a issue #9784 that was reported a few months ago regarding this window ([Changing company color scheme is confusing](https://github.com/OpenTTD/OpenTTD/issues/9784)). In the presented mock-up in this form, this problem will not only not be solved, but will even be slightly strengthened. As adding this function would involve some rebuilding of the window, I think it would be a good idea to make this rebuilding in such a way that it would also improve the way of selection.

I admit that for me the current way of choosing colors is also often confusing. Especially when there are many groups and there is a need to change many colors. It just asks you to be able to change the color by pressing directly where it is marked. In the pictures below, I present you some ideas...

![Special scheme 2 1](https://user-images.githubusercontent.com/67662046/171257382-ede1bc14-ea2d-477d-ac0a-305abcd3508c.png)

**1. Choose a color by pressing directly on the name**
![Special scheme 2 1c](https://user-images.githubusercontent.com/67662046/171260525-14c31a95-7cc3-43bf-87ec-acd373effbf2.png)
You don't have to select a line first, then open the list in another part and then make a selection. Here you click on the part what you want to change and a selection box opens below or above the line. The number of steps you have to do is thus kept to a minimum, and the place you have to click to change the color is logical and intuitive. In the upper part, instead of the selection bar, there are painting type titles. It is important that the painting selection window has a scroll bar as this list can be really long. Instead of small icons with colors, there could be any other graphics defined by NewGRF. It would be good if the paintings that are not available for all vehicles in the group should be marked with a less distinct color.

**2. Adding a warning icon in case not all vehicles in the group can use the selected painting.**
![Special scheme 2 2](https://user-images.githubusercontent.com/67662046/171258724-dfedb95f-c887-4824-b1c1-1c18877935bc.png)
I think that a significant problem may be the situation when not all vehicles in the group, and sometimes almost none, will be able to use the selected painting. For an experienced player it will not be a problem, but for someone new it will be a irritating bug. _"Why don't the vehicles have BRBLUE livery, although I have set one like that?! Why only one has? WTF!"_ This problem can be partially solved by adding a warning icon next to each group, where not all vehicles have such a scheme available. Also an explanation should appear at the bottom of the window when a player clicks on a problematic group.

**3. If none of the vehicles has any special painting, it should be marked accordingly.**
![Special scheme 2 5](https://user-images.githubusercontent.com/67662046/171261752-35a87de1-397d-41b1-9ac8-6151aa58febb.png)
Faded Gray seems to be the best solution to me as it is an unavailable option, so it shouldn't distract the player. By clicking on a field, an appropriate explanation could appear in the bottom line, but in my opinion it is not necessary. The painting icon, in the case of unavailability and no choice, could be empty ("default" can have two colors).

**4. If the group is assigned a special painting, it could be marked with a distinctive color.**
![Special scheme 2 4](https://user-images.githubusercontent.com/67662046/171262299-8139bcd6-2caf-45c4-8958-ab48f8e5e93b.png)
This is because the special painting is dominant over the standard colors and if selected, changing the other colors has no effect. Thanks to the distinctive color of the names, this behavior seems more logical than when all the painting are of equal level.

**5. Highlight painting when the cursor is over a field.**
![Special scheme 2 3b](https://user-images.githubusercontent.com/67662046/171263319-51b9b21b-050d-4b4a-a7bd-e00f925a9357.png)
This makes it easier to see which setting you want to change. Currently, it would be inconsistent with the rest of the interface. But I show this possibility, because perhaps it would be worth taking this path. The text of the interactive element would be highlighted and if clicked, it would trigger some reaction. This would provide new possibilities, especially where adding a button is troublesome (eg in the finance window, clicking on the text "Trains" in the revenue group would open a window showing more information - company's trains grouped by profit last year). The change would have to apply to almost all windows, but perhaps it would be possible to do it globally(?).

**Reply by FranticFairy** [reply]
Thank you for your input! While I do agree with some points, there are a few that, at least in my opinion, would not make sense to also incorporate into this potential change, which will be the ones I'll adress below

3 - I do not think it should behave this way. Instead, I think you should be able to set a livery regardless of whether or not any vehicles in the group can make use of it. This is so that you can set a Group Livery in advance - setting BR Blue in preparation of that livery becoming available, for example - and to make sure there will be no weirdness in case the livery becomes 'unavailable' for the group later on through, for example, changing to different trains.

4 - I have somewhat mixed feelings about this. While adding a distinctive color to Group Liveries could be nice, there is no guarantee that changing the default colours has no effect - a GRF developer could make use of Group Liveries to give a different 2CC sprite to vehicles, in which case setting CC1 and CC2 would still affet the vehicles in the group.

As for points 1 and 5, I do agree that they could be convenient changes, however I also believe the UI should remain consistent in terms of behaviour compared to other screens. As such, I don't believe this functionality should be changed purely because "It would be more convenient" without also changing the behaviour of other screens, and incorporating an entire UI rewrite/change/however you want to call it would be far, far out of scope of this suggestion.

And finally, point 2. I can get behind something like this being included in some way. It's a good feature to have, and would avoid confusion among newer users.

**Reply by FranticFairy** [reply]
Actually, upon further discussion, I've realized that point 2 might not be viable, unless vehicles have to declare in advance every livery they support.

**Reply by LC-Zorg** [reply]
> 3 - I do not think it should behave this way. Instead, I think you should be able to set a livery regardless of whether or not any vehicles in the group can make use of it. This is so that you can set a Group Livery in advance - setting BR Blue in preparation of that livery becoming available, for example - and to make sure there will be no weirdness in case the livery becomes 'unavailable' for the group later on through, for example, changing to different trains.

For my use and most of what I watched, this seemed like a good solution, but I agree with what you write. But... take a look also at the fact that even one NewGRF can have dozens of different paintings. If you combine several add-ons, the number of its may exceed 100. And it is very likely that the vehicles in the group will not be able to use any of the settings. The player will set one setting after the other and all the time it will have no effect or the effect will be only partial. I believe that you shouldn't assume that all or even most players know what they are doing and that they know all the possible paintings, or that they only use add-ons from one author who has thought and refined everything. It would be useful some information otherwise it will be very frustrating.

This could be something similar to what I showed in the pick list - lighter text color means less matching of the scheme to the current vehicles in the group.
![Show match v2](https://user-images.githubusercontent.com/67662046/171292705-79b4f4c5-b4fd-4286-9ada-8717ebfc817b.png)
Above, I used different shades (I don't know if it's possible now) and the percentage indicator - player choosing to paint will see what to expect. Instead of a warning icon, you can also use the percentage indicator (none if 100%) for painting names.

Another, separate solution could be the option to filter vehicles in the purchase and replacement window, where you could also filter by selecting paints.

Another solution could be the painting catalog...

> 4 - I have somewhat mixed feelings about this. While adding a distinctive color to Group Liveries could be nice, there is no guarantee that changing the default colours has no effect - a GRF developer could make use of Group Liveries to give a different 2CC sprite to vehicles, in which case setting CC1 and CC2 would still affet the vehicles in the group.

The fact is, I did not think about this that the paint scheme can define the scheme itself without the colors. :)

> As for points 1 and 5, I do agree that they could be convenient changes, however I also believe the UI should remain consistent in terms of behaviour compared to other screens. As such, I don't believe this functionality should be changed purely because "It would be more convenient" without also changing the behaviour of other screens, and incorporating an entire UI rewrite/change/however you want to call it would be far, far out of scope of this suggestion.

Point 5 actually differs significantly from what is known in the game and the change would have to cover everything to be consistent. I don't know what is the reality of this idea. This is just a casual proposition. :)

As for point 1, I also made versions with multiple picklists before, which would be more consistent. Currently something similar is used in the JGRPP version for selecting loading / unloading orders by cargoes. Well... it looks so-so, although I think that it would be a much better solution than the current one anyway. It would also solve the mentioned problem.
![Special scheme 3 1](https://user-images.githubusercontent.com/67662046/171293959-149d73a2-0b1a-4003-9a6f-9df6c428b57d.png)

**Reply by FranticFairy** [reply]
Just chiming in quickly to elaborate on points 2 & 3 and why they're... Complicated, based on what I understood after talking about it with people much more familiar with the inner workings of OpenTTD than me.

Unless a special static property or something is implemented, vehicles have no way of telling the game what liveries they support - their livery functionality would just be "I read what value I have for livery, and do stuff based on that".

Because vehicles can't tell the game what liveries they support, the game can't know how many vehicles ik a group will be affected by the game. I do get why something like this would be convenient to have, but I do not know how viable it is to implement support for anything like this without sacrificing other functionality.

**Reply by LC-Zorg** [reply]
Well, ad nauseam (or to change status) I can only say that I am not a programmer and my ideas are just from fantasy about how this might work. ;) However, I am wondering how it is that the game has to know what are all available schemes and make them available to the player to choose from, but at the same time it cannot determine what schemes the vehicles in the group have. I look at it from the player's point of view, and here, after selecting the vehicle, it is possible to determine what schemes the vehicle has. So why can't the game open it virtually? Since there is a way to display information about available schemes, there should probably be a way to access that information as well. Based on them, the game would create a static list. Of course, all of this could be bul...
Another concept is actually like you wrote to create in each (new) addon a list of available schemes that the game could easily read (such a list could also solve other problems such as the one related to capacities #9584). This means that for older add-ons the feature would not work, but for newer ones it would be possible to do it in a way that would not cause confusion and frustration among players.
I really wouldn't want to break this, but I believe that this is a significant problem of this feature which not should be ignored. I believe that adding a function to the game that would not work even in 100% of cases or work fragmentarily. in a way incomprehensible to players, would be a bad thing.

**Reply by LC-Zorg** [reply]
BTW, I wonder if such a list could not also be a way of entering vehicle data in add-ons...

**Comment by andythenorth** [comment]
Alternative idea.

Option to save named combos of company colours. 

Groups would default to “default”. Another option would be "custom". Another option would be choosing a saved combo.

A mockup would explain this better but I can't make one currently 😜

**Comment by andythenorth** [comment]
For now, we have vehicle Variants, via #10220 and these will be in 13.0

Whilst it might be nice in future to have a specific livery selection mechanism (skins, or combos of company colours), let's see how variants shakes out first 😄  - usage of this type of feature takes a while to bed in, and designing further features while that's happening usually doesn't go so well 😄
